### PR TITLE
Gazelle select (8/x): methods for adding files to Package

### PR DESCRIPTION
**Pull request by jayconrod**
Related #409

### CODE REVIEW

File: go/tools/gazelle/packages/package.go [hunk 1]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+// "buildTags" is a set of tags that are true on all platforms.
+// "platforms" is a map from platform names (labels referencing config_settings)
+// to sets of tags that are true on those platforms. The tag sets may include
+// "buildTags".
+//
+// An error is returned if a file is buildable but invalid (for example, a
+// test .go file containing cgo code). Files that are not buildable will not
+// be added to any target (for example, .txt files).
+func (p *Package) addFile(info fileInfo, cgo bool, buildTags map[string]bool, platforms PlatformConstraints) error {
+	var t *Target
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: I feel like this function would be easier to read as
```
	switch {
	case info.isXTest:
		if info.isCgo {
			return fmt.Errorf("%s: use of cgo in test not supported", info.path)
		}
		p.XTest.addFile(info, buildTags, platforms)
	case info.isTest:
		if info.isCgo {
			return fmt.Errorf("%s: use of cgo in test not supported", info.path)
		}
		p.Test.addFile(info, buildTags, platforms)
	case info.isCgo || cgo && (info.category == cExt || info.category == hExt || info.category == csExt):
		p.CgoLibrary.addFile(info, buildTags, platforms)
	case info.category == goExt || info.category == sExt || !cgo && info.category == hExt:
		p.Library.addFile(info, buildTags, platforms)
	}
	return nil
```

File: go/tools/gazelle/packages/package.go [hunk 2]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+		}
+		t = &p.XTest
+	} else if info.isTest {
+		if info.isCgo {
+			return fmt.Errorf("%s: use of cgo in test not supported", info.path)
+		}
+		t = &p.Test
+	} else if info.isCgo || cgo && (info.category == cExt || info.category == hExt || info.category == csExt) {
+		t = &p.CgoLibrary
+	} else if info.category == goExt || info.category == sExt || !cgo && info.category == hExt {
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: isn't the !cgo redundant, it will have been matched in the condition above if true?
(I did not fix this in my suggested switch version above btw)

File: go/tools/gazelle/packages/package.go [hunk 3]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+func (ps *PlatformStrings) Clean() {
+	sort.Strings(ps.Generic)
+	ps.Generic = uniq(ps.Generic)
+
+	genSet := make(map[string]bool)
+	for _, s := range ps.Generic {
+		genSet[s] = true
+	}
+
+	if ps.Platform != nil {
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: Invert the condition to a return and you can undent the rest of the function

File: go/tools/gazelle/packages/package.go [hunk 4]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+		for n, ss := range ps.Platform {
+			ss = remove(ss, genSet)
+			if len(ss) == 0 {
+				delete(ps.Platform, n)
+				continue
+			}
+			sort.Strings(ss)
+			ps.Platform[n] = uniq(ss)
+		}
+		if len(ps.Platform) == 0 {
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: Why is this necessary? Does something really care about Platform == nil rather than len(Platform) == 0? if so, it probably should not.

File: go/tools/gazelle/packages/package.go [hunk 5]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+	ps.Generic = uniq(ps.Generic)
+
+	genSet := make(map[string]bool)
+	for _, s := range ps.Generic {
+		genSet[s] = true
+	}
+
+	if ps.Platform != nil {
+		for n, ss := range ps.Platform {
+			ss = remove(ss, genSet)
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: I feel like this whole bit would be much easier to understand (and probably faster) if you just converted it to a map, then twiddled the bools (possibly converting back to a list at the end if you really really need a list rather than a map), rather than sorting and uniquing

File: go/tools/gazelle/packages/package.go [hunk 6]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+		if !remove[ss[r]] {
+			ss[w] = ss[r]
+			w++
+		}
+	}
+	return ss[:w]
+}
+
+func uniq(ss []string) []string {
+	if len(ss) <= 1 {
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: pointless precondition? first test of the loop does this anyway

File: go/tools/gazelle/packages/package.go [hunk 7]
```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
...
+		return ss
+	}
+
+	var r, w int
```
Comment (ianthehat) [inline go/tools/gazelle/packages/package.go]: only w is used outside the loop, and r is the true loop control variable, so you might as well declare and init (with :=) w outside and r inside
also you can compare ss[r] to ss[w] becuase r-1 will have been copied down there anyway
also, r is used as a straight up iterator...
also, you might as well just cheat and use append rather than count and reslice, it's not going to be much worse in performance and much easier to understand
```
func uniq(ss []string) []string {
	result := ss[:0]
	previous := ""
	for _, s := range ss {
		if s != previous {
			result = append(result, s)
			previous = s
		}
	}
	return result
}
```

File: go/tools/gazelle/packages/package.go (hunk 1 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: Done

File: go/tools/gazelle/packages/package.go (hunk 2 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: Yes, it is redundant. Thought it would make the condition more readable, but maybe it's just confusing.

File: go/tools/gazelle/packages/package.go (hunk 3 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: Done

File: go/tools/gazelle/packages/package.go (hunk 5 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: It's important that this stays a list that normally allows duplicates. The sorting and uniquing isn't important for the final result (buildifier does this on the AST), but it's useful to have predictable results for unit testing.

`PlatformStrings` is used for files, flags, and imports. `Clean` doesn't need to be called on files or flags. Files are only added once in a predictable order. Flags are not sorted, and duplicates are allowed.

`Clean` does need to be called on imports. Imports can be added repeatedly if they appear in different files, and they can be both generic and platform-specific. It should be possible to prevent duplication and to move platform-specific imports to the generic list if they are added to the generic list later, but it would require some extra internal data structures in `PlatformStrings` that would only be used for imports. It seemed simpler to `Clean` after everything is added.

File: go/tools/gazelle/packages/package.go (hunk 4 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: Nothing cares about this, but it made the tests a little easier to write (don't need to create an empty map), and it saves a little memory.

File: go/tools/gazelle/packages/package.go (hunk 7 above)
Comment (jayconrod) [inline go/tools/gazelle/packages/package.go]: I've mostly taken your proposed code. I think I need that precondition again now though.
